### ISSUE TITLE
Strange behavior with exceptions in flatMapConcat

### ISSUE THREAD

**Issue by iravid** [akka/akka#21191]
Hello, I've noticed some strange behaviors while working with failures inside flatMapConcat.

Folowing are a few reproductions to illustrate.

First, this simple stream correctly reports an error in `graphDecider`: 
```scala
implicit val system = ActorSystem("stream-tests")
    
val graphDecider: Supervision.Decider = { e =>
  println(s"Caught ${ e } in decider1; stopping")
  Supervision.Stop
}

val actorMatDecider: Supervision.Decider = { e =>
  println(s"Caught ${ e } in decider2; stopping")
  Supervision.Stop
}

implicit val mat = ActorMaterializer(ActorMaterializerSettings(system).withSupervisionStrategy(decider2))

val stream = Source
  .single("hello")
  .map(_ => throw new Exception("inner"))
  .to(Sink.foreach(println))
  .withAttributes(ActorAttributes.supervisionStrategy(decider1))

stream.run()
```

However, materializing a source with the same `map` stage inside flatMapConcat catches the error in the `matDecider`:

```scala
val stream = Source
  .single("hello")
  .flatMapConcat { s =>
    Source.single(s).map(_ => throw new Exception("inner"))
  }
  .to(Sink.foreach(println))
  .withAttributes(ActorAttributes.supervisionStrategy(graphDecider))
```

Even stranger, materializing a `Source.failed` simply hangs:
```scala
val stream = Source
  .single("hello")
  .flatMapConcat { s =>
    Source.failed(new Exception("inner"))
  }
  .to(Sink.foreach(println))
  .withAttributes(ActorAttributes.supervisionStrategy(graphDecider))
```

The behavior I would expect (and please correct me if I missed something in the docs :-)) is:
1. A failed source in `flatMapConcat` should be treated as though an exception was thrown and should complete the stage with an error
2. A failed stage in a source materialized in `flatMapConcat` should be caught in the graph's decider

**Comment by drewhk** [comment]
> However, materializing a source with the same map stage inside flatMapConcat catches the error in the matDecider:

The reason for this is that the attribute of the Source created inside the flatMapConcat is not inherited from the stage which creates it. It inherits directly from the materializer. In other words, attributes are statically assigned, we don't check if a stream was started inside another stream and take attributes from it. This is by design and changing it would be horribly complicated.

> Even stranger, materializing a Source.failed simply hangs:

That is likely a bug though.

**Comment by iravid** [comment]
Thanks for the answer! Makes sense once you stop considering stream deciders as hierarchical in any way :-)

**Comment by drewhk** [comment]
The flatMapConcat with the failed source is likely a bug though, so we need to still investigate that.

**Comment by drewhk** [comment]
Er, the flatMapConcat example works with the failed source. The Future that is materialized by the Sink.foreach contains the exception properly. It does not hang at all for me.

**Comment by hderms** [comment]
@drewhk so if the source passed to flatMapConcat fails then the outer supervisor isn't triggered? This behavior seems pretty strange and unintuitive to me and is causing a bug in our production system.

**Comment by hderms** [comment]
@drewhk I think there were a few points @iravid made that haven't been addressed and really, the absence of this behavior makes no sense to me:

1. A failed source in flatMapConcat should be treated as though an exception was thrown and should complete the stage with an error
1. A failed stage in a source materialized in flatMapConcat should be caught in the graph's decider

is there any explanation for why a failed source doesn't bubble up to the decider?

**Comment by drewhk** [comment]
@hderms I think you misunderstand streams supervision. It is a very limited abstraction compared actors supervision. Namely:
 - it is either supported by a stage or not as it has to be explicitly built into stages
 - it is not hierarchical, only applied to single stages. This means there is no bubbling up.

The difference in the two cases by @iravid :

Case 1:
```scala
val stream = Source
  .single("hello")
  .flatMapConcat { s =>
    Source.single(s).map(_ => throw new Exception("inner"))
  }
  .to(Sink.foreach(println))
  .withAttributes(ActorAttributes.supervisionStrategy(graphDecider))
```
In this case the supervision set on the materializer is applied to the `map` in question. It handles the exception and recovers, causing the element to be simply dropped. So the `Source.single()` simply turns into an empty Source (not a failed one)

Case 2:
```scala
val stream = Source
  .single("hello")
  .flatMapConcat { s =>
    Source.failed(new Exception("inner"))
  }
  .to(Sink.foreach(println))
  .withAttributes(ActorAttributes.supervisionStrategy(graphDecider))
```
The inner source fails, which in turn fails the `flatMapConcat`, as it does not have supervision support. 

If you want to simply drop failed sources in `flatMapConcat` then you can use the `recover` combinator on the inner sources.

**Comment by hderms** [comment]
@drewhk that's what we ended up doing but it's confusing that there's no way to make the failed source bubble up into the supervisor in the second case. We want to be able to log exceptions and we were previously doing that with supervisors, however, we now will have to use multiple approaches to log exceptions because this behavior seemingly differs from the rest of akka streams. 
I understand that we could ensure that any flatMapConcat has a recover (and as part of the recover combinator we log the exception) but it was definitely countintuitive to have our stream stop processing messages because of a seemingly benign failure. We are trying to figure out a way to make this more resilient as that was our entire reason for using Akka.  

Perhaps we are doing something wrong. We are consuming messages from rabbitmq as the source of a stream (using op-rabbit) and part of the processing of this message uses a flatMapConcat to paginate through results (flatMapConcat with an inner source using unfoldAsync) so that each page gets spit out from the flatMapConcat. This pattern seemed elegant at first until we ran into the flatMapConcat issue. 

Should we be instantiating a new stream for each rabbit message and then executing it? This would prevent a failed stream from accidentally failing the entire stream (and therefore preventing further messages from being processed via rabbitmq). 

Optionally we could also just restart the entire process when the stream fails which is problematic for its own reasons. What we really want is for the flatMapConcat to be restarted the same as any other stage. 

What are your opinions on this as we would prefer not to have to ensure every inner stream of a flatMapConcat has to use `recover` in some particular way to log exceptions.

**Comment by drewhk** [comment]
> that's what we ended up doing but it's confusing that there's no way to make the failed source bubble up into the supervisor in the second case.

Remember that flatMapConcat is not necessarily constructing the sources themselves (maybe they come from an upstream `splitWhen`). The function provided to flatMapConcat could actually extract a source from a received element. I.e. there is no necessary parent-child relationship between the concat stage and the flattened sources per se. In those cases, who should supervise the Sources?. In fact, these sources can be wrapped, raw Reactive Streams Publishers, in which case there is no hope for supervising them in any way as the RS protocol does not specify any means for supervision. In other words, the _only_ reliable way to see if a Source failed is by observing it and see if an error bubbles up as a source failure. Restarting a source is impossible for flatMapConcat itself (but you might be able to use `recoverWith` though).

> We are trying to figure out a way to make this more resilient as that was our entire reason for using Akka.

Streams supervision, as you would like it to see, is an _extremely_ hard problem. There is a reason why it is not implemented in full in Akka Streams. From all libraries in the Reactive Streams space, basically only Akka has the potential to implement more powerful supervision (but even then, with limitations), but it is very hard work and the full solution is not yet known. I would imagine that if it is solved, it would be a commercial add-on even, given that it is such a valuable and hard to solve design problem.

In other words, I hear your problem, and I agree it would be nice if it would "just work" as you imagine, but it was not implemented in the current, limited way due to ignorance, but because of the nature of the problem.

> What are your opinions on this as we would prefer not to have to ensure every inner stream of a flatMapConcat has to use recover in some particular way to log exceptions.

You only need to add the `recover` or `recoverWith` stages in the function you provide to flatMapConcat, the upstream sources do not need to know about it. I.e. you can centralize this addition to that single point.
